src/UglyToad.PdfPig/DocumentLayoutAnalysis/Distances.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Linq;
+using UglyToad.PdfPig.Geometry;
+
+namespace UglyToad.PdfPig.DocumentLayoutAnalysis
+{
+    /// <summary>
+    /// Contains helpful tools for distance measures.
+    /// </summary>
+    public static class Distances
+    {
+        /// <summary>
+        /// The Euclidean distance is the "ordinary" straight-line distance between two points.
+        /// </summary>
+        /// <param name="point1">The first point.</param>
+        /// <param name="point2">The second point.</param>
+        /// <returns></returns>
+        public static double Euclidean(PdfPoint point1, PdfPoint point2)
+        {
+            double dx = (double)(point1.X - point2.X);
+            double dy = (double)(point1.Y - point2.Y);
+            return Math.Sqrt(dx * dx + dy * dy);
+        }
+
+        /// <summary>
+        /// The weighted Euclidean distance.
+        /// </summary>
+        /// <param name="point1">The first point.</param>
+        /// <param name="point2">The second point.</param>
+        /// <param name="wX">The weight of the X coordinates. Default is 1.</param>
+        /// <param name="wY">The weight of the Y coordinates. Default is 1.</param>
+        /// <returns></returns>
+        public static double WghtdEuclidean(PdfPoint point1, PdfPoint point2, double wX = 1.0, double wY = 1.0)
+        {
+            double dx = (double)(point1.X - point2.X);
+            double dy = (double)(point1.Y - point2.Y);
+            return Math.Sqrt(wX * dx * dx + wY * dy * dy);
+        }
+
+        /// <summary>
+        /// The Manhattan distance between two points is the sum of the absolute differences of their Cartesian coordinates.
+        /// <para>Also known as rectilinear distance, L1 distance, L1 norm, snake distance, city block distance, taxicab metric.</para>
+        /// </summary>
+        /// <param name="point1">The first point.</param>
+        /// <param name="point2">The second point.</param>
+        /// <returns></returns>
+        public static double Manhattan(PdfPoint point1, PdfPoint point2)
+        {
+            return (double)(Math.Abs(point1.X - point2.X) + Math.Abs(point1.Y - point2.Y));
+        }
+
+        /// <summary>
+        /// Find the nearest point.
+        /// </summary>
+        /// <param name="pdfPoint">The reference point, for which to find the nearest neighbour.</param>
+        /// <param name="points">The list of neighbours candidates.</param>
+        /// <param name="measure">The distance measure to use.</param>
+        /// <param name="dist">The distance between reference point, and its nearest neighbour</param>
+        /// <returns></returns>
+        public static PdfPoint FindNearest(this PdfPoint pdfPoint, PdfPoint[] points,
+            Func<PdfPoint, PdfPoint, double> measure, out double dist)
+        {
+            double d = points.Min(k => measure(k, pdfPoint));
```

Comment:
Might be worth a few precondition checks here with detailed exceptions, one to check `points` and `measure` (maybe `distanceCalculator` is a better name?) for null and one to check `points` is not empty (in which case the call to min fails but the exception from the Linq could always be more detailed :'( ).

Reply:
On performance I think I'd prefer the following:

```suggestion
            var minDistance = double.MaxValue;
            Point closestPoint = default;
            for (var i = 0; i < points.Length; i++)
            {
                var distance = measure(points[i], pdfPoint);
                if (distance < minDistance) 
                {
                   minDistance = distance;
                   closestPoint = points[i];
                }
            }
            
            return closestPoint;
```

Where `points.Length` is `points.Count` for the `IReadOnlyList<PdfPoint>` case. This prevents needing to iterate over the collection twice with a call to `measure` on each iteration.